Python migration guide (#3634)
chainguard-dev/edu: content/chainguard/libraries/python/migration.md

* [chainctl installed and authenticated](/chainguard/chainctl-usage/how-to-install-chainctl/)
* An [entitlement to Chainguard Libraries](/chainguard/chainctl/chainctl-docs/chainctl_libraries_entitlements_create/) for Python

### Create an entitlement

To create an entitlement to Chainguard Libraries for Python and enable upstream fallback, run:

```shell
chainctl libraries entitlements create --ecosystems=PYTHON --policy=CHAINGUARD_AND_UPSTREAM
```

Alternatively, you can create an entitlement and pull token in the Chainguard Console: while viewing the Python ecosystem page, follow the prompts to create an access token.

It can take up to 30 minutes for fallback policy changes to take effect.

### Authentication prerequisites

For authentication, you need a pull token or the Python keyring provider.

{{< tabs >}}

{{% tab title="Pull token" %}}

If you plan to use a repository manager, or a non-interactive environment such as CI/CD, you will need a pull token. You must be an Owner or have the `libraries.python.pull_token_creator` permission to create one.

```shell

> chainctl auth pull-token --repository=python --name=my-python-token --ttl=720h

Reviewer: i think the more explicit `chainctl auth pull-token create` would be preferable here to `chainctl auth pull-token`.  ie, be explicit on `create` sub-command as there is also a `list` subcommand.

pip:

```bash
pip-compile --generate-hashes --index-url https://libraries.cgr.dev/python/simple/ requirements.in
```

uv:

```bash
rm -f uv.lock  
uv sync
```

Poetry 1.x:

```bash
poetry lock --no-update
```

Poetry 2.x:

```bash
poetry lock
```

{{< /details >}}

## Step 3: Clear caches

{{< tabs >}}

{{% tab title="pip" %}}

```shell
pip cache purge
```

{{% /tab %}}

{{% tab title="uv" %}}

```bash
uv cache clean
```

{{% /tab %}}

{{% tab title="Poetry" %}}

```bash
poetry cache clear --all pypi
```

{{% /tab %}}

{{< /tabs >}}

Also invalidate cached packages in any repository manager that previously proxied PyPI, and rebuild containers without cached layers if your Dockerfile installs dependencies during build:

```shell
docker build --no-cache .
```

## Step 4: Reinstall dependencies

Reinstall dependencies and confirm that the lockfile reflects Chainguard as the source.

{{< tabs >}}

{{% tab title="pip" %}}

```shell
pip install --force-reinstall --no-cache-dir -r requirements.txt
```

{{% /tab %}}

{{% tab title="uv" %}}

Delete the virtual environment outright and let uv rebuild it from the current lockfile:

```bash
rm -rf .venv
uv sync

> ```

Reviewer 2: suggestion: I believe `uv sync --force-reinstall` should do the same without needing to be aware of where the virtualenv actually is

Author: Thanks! I just tested this again and it worked with `uv sync --reinstall`

tags: ["Chainguard Libraries", "Python"]
menu:
  docs:
    parent: python
    identifier: Python Migration
weight: 056
toc: true
---

Chainguard Libraries for Python provides a curated index of PyPI packages rebuilt from source and verified against Chainguard's build provenance. Because Chainguard Libraries uses the standard pip simple-index protocol, switching an existing project requires only an index configuration change — no changes to your application code or dependency declarations.

This guide walks through migrating an existing Python project to Chainguard Libraries, covering the two most common setups:

* **Direct access**: Your build tool connects directly to `libraries.cgr.dev`. This option is faster for initial evaluation and smaller-scale setups.  
* **Repository manager**: Your build tool connects to a repository manager (such as JFrog Artifactory or Sonatype Nexus), which proxies requests to Chainguard Libraries. This option is recommended for teams and organizations.

## Prerequisites

Before you begin, you need:

* An existing, working Python project with a `requirements.txt`, `poetry.lock`, `uv.lock`, `pdm.lock`, `Pipfile.lock`, or `pylock.toml`

> * [A pull token](#authentication-prerequisites) or the [Python keyring provider](/chainguard/libraries/access/#python-keyring-provider)

Reviewer 3: This is a little off, as the guide shows how to create a pull token below, making this not exactly a 'prerequisite'. Maybe reframe this by taking away this bullet, and following the list of prereqs with something like "you'll also need either a pull token or the python keyring provider" and then give the instructions there

    * The keyring provider requires pip 23.1 or later. Older pip versions cannot automatically discover or invoke a keyring backend.
* [chainctl installed and authenticated](/chainguard/chainctl-usage/how-to-install-chainctl/)

> * An [entitlement to Chainguard Libraries](/chainguard/chainctl/chainctl-docs/chainctl_libraries_entitlements_create/) for Python

Reviewer 3: similarly, I think it's different from needing an existing python project or chainctl installed since you explain how to create the entitlement right after this

Author: I reorganized this section, lmk what you think!

### Create an entitlement

To create an entitlement to Chainguard Libraries for Python and enable upstream fallback, run:

```shell
chainctl libraries entitlements create --ecosystems=PYTHON --policy=CHAINGUARD_AND_UPSTREAM
```

It can take up to 30 minutes for fallback policy changes to take effect.

### Authentication prerequisites

For authentication, you need a pull token or the Python keyring provider.

{{< tabs >}}

{{% tab title="Pull token" %}}

> If you plan to use a repository manager, or a non-interactive environment such as CI/CD, you will need a pull token. You must be an Owner or have the `libraries.python.pull_token_creator` permission to create one.

Reviewer 3: ```suggestion
If you plan to use a repository manager, or a non-interactive environment such as CI/CD, you will need a pull token. You must have the `owner` role or the `libraries.python.pull_token_creator` capability to create one.
```

```shell
chainctl auth pull-token create --repository=python --name=my-python-token --ttl=720h
```

To export environment variables directly:

```shell
eval $(chainctl auth pull-token --output env --repository=python --name=my-python-token)
```

Learn more about creating and managing pull tokens in the [Libraries Access documentation](/chainguard/libraries/access/#creating-pull-tokens-for-libraries).

{{% /tab %}}

{{% tab title="Python keyring" %}}

The keyring leverages `chainctl` to fetch temporary credentials whenever your environment requests packages from Chainguard. Supported environments include local development and CI/CD platforms that can use assumable identities.

Learn how to install this package in the [Chainguard Libraries Access documentation](chainguard/libraries/access/#python-keyring-provider).

Note that `.netrc` takes precedence over the keyring. Verify that you do not have an existing registry entry in `.netrc` for Chainguard Libraries:

```shell
grep -A2 "libraries.cgr.dev" ~/.netrc
```

Remove or correct any stale entry before proceeding.

{{% /tab %}}

{{< /tabs >}}

> > **Note: Do not commit credentials to version control.** Store tokens as environment variables or CI secrets, not as literal values in `pip.conf`, `.netrc`, or any file that might be checked into a repository.

Reviewer 3: love this note!

Once configured, point your build tool at your repository manager URL instead of `libraries.cgr.dev` directly. In this setup, the credentials are your repository manager credentials — not a Chainguard pull token.

{{< tabs >}}

{{% tab title="pip" %}}

```shell
pip config set global.index-url https://<your-repo-manager-url>/repository/python-all/simple/
```

{{% /tab %}}

{{% tab title="uv" %}}

Configure `~/.config/uv/uv.toml` for a global default across all projects:

```bash
[[tool.uv.index]]
   name = "chainguard"
   url = "https://<your-repo-manager-url>/repository/python-all/simple/"
   default = true
```

For a per-project configuration, configure the `pyproject.toml`:

```pyproject.toml
[[tool.uv.index]]
name = "python-all"
url = "https://repo.example.com/repository/python-all/simple/"
```

{{% /tab %}}

{{% tab title="Poetry" %}}

Run the following command:

```shell
poetry config http-basic.python-all REPO_MANAGER_USERNAME REPO_MANAGER_PASSWORD
poetry source add python-all https://repo.example.com/repository/python-all/simple/
```

{{% /tab %}}

{{< /tabs >}}

Example URLs by repository manager:

> | Repository manager | URL pattern |

Reviewer 3: very nice table!

| ----- | ----- |
| JFrog Artifactory | `https://example.jfrog.io/artifactory/api/pypi/python-all/simple/` |
| Sonatype Nexus | `https://repo.example.com:8443/repository/python-all/simple/` |
| Google Artifact Registry | `https://<location>-python.pkg.dev/<project>/python-all/simple/` |
| Cloudsmith | `https://dl.cloudsmith.io/.../<org>/python-all/python/simple/` |

> For authentication details specific to your repository manager, see the [global configuration documentation](/chainguard/libraries/python/global-configuration/).

Reviewer 3: ```suggestion
For authentication details specific to your repository manager, refer to the [global configuration documentation](/chainguard/libraries/python/global-configuration/).
```

## Step 2: Update your lockfile

> Your existing lockfile or hash-pinned requirements.txt contains checksums generated against packages downloaded from PyPI or through your repository manager. Because Chainguard rebuilds packages from verified source as well as providing security controls for upstream artifacts, checksums differ even for identical version numbers. If your file uses `--hash` entries or `--require-hashes`, installation will fail with a hash mismatch after switching to Chainguard until these are updated.

Reviewer 3: ```suggestion
Your existing lockfile or hash-pinned `requirements.txt` contains checksums generated against packages downloaded from PyPI or through your repository manager. Because Chainguard rebuilds packages from verified source as well as providing security controls for upstream artifacts, checksums differ even for identical version numbers. If your file uses `--hash` entries or `--require-hashes`, installation will fail with a hash mismatch after switching to Chainguard until these are updated.
```

this aligns with the styling in the prerequisites

Run the following command to auto-detect and update the lockfile in the current project:

```bash
chainctl libraries update-hashes
```

Or specify the lockfile when running the command. For example:

```bash
chainctl libraries update-hashes requirements.txt
```

When using a repo manager, run a command similar to the following:

```bash
chainctl libraries update-hashes
  --registry-url https://repo.example.com/repository/python-all/
  --token "$REPO_TOKEN"
```

By default, Chainguard hashes are appended alongside your existing hashes rather than replacing them. If your installer fails on a dual-hash entry, use the `--replace` flag.

### Alternative: Regenerate the lockfile

Regenerating is **not recommended** as a first approach. Dependency resolvers pick the newest version satisfying your constraints, which may not yet be available from Chainguard, and may also introduce unrelated dependency upgrades.

> {{< details "Delete and regenerate a lockfile" >}}

Reviewer 3: I'm not sure about how you've formatted this section. Like, the way it looks in the preview is like we made a mistake trying to get tabs. It's just kind of jarring, and looks almost unfinished compared to other sections.

If you have a specific reason to regenerate - for example, if you are intentionally refreshing your dependency versions -  use the following commands:

pip:

```bash
pip-compile --generate-hashes --index-url https://libraries.cgr.dev/python/simple/ requirements.in
```

uv:

```bash
rm -f uv.lock  
uv sync
```

Poetry 1.x:

```bash
poetry lock --no-update
```

Poetry 2.x:

```bash
poetry lock
```

{{< /details >}}

## Step 3: Clear caches

> 

Reviewer 3: Ideally there'd be a sentence here for a little added context. I'm sure it would just reiterate the heading but maybe just drive home why we need to clear caches

{{< tabs >}}

{{% tab title="pip" %}}

```shell
pip cache purge
```

{{% /tab %}}

{{% tab title="uv" %}}

```bash
uv cache clean
```

{{% /tab %}}

{{% tab title="Poetry" %}}

```bash
poetry cache clear --all pypi
```

{{% /tab %}}

{{< /tabs >}}

Also invalidate cached packages in any repository manager that previously proxied PyPI, and rebuild containers without cached layers if your Dockerfile installs dependencies during build:

```shell
docker build --no-cache .
```

## Step 4: Reinstall dependencies

> Reinstall dependencies and confirm that the lockfile reflects Chainguard as the source.

Reviewer 3: (following comment immediately above this)

this context is, if nothing else, a little reassuring for the reader. I think it's also nicer visually to go from heading -> sentence -> command than heading -> command

{{< tabs >}}

{{% tab title="pip" %}}

```shell
pip install --force-reinstall --no-cache-dir -r requirements.txt
```

{{% /tab %}}

{{% tab title="uv" %}}

```bash
uv sync --reinstall
```

{{% /tab %}}

{{% tab title="Poetry" %}}

```bash
poetry env remove --all
poetry install
```

{{% /tab %}}

{{< /tabs >}}

## Step 5: Verify your libraries

After reinstalling, you can use `chainctl` to verify which dependencies are built by Chainguard. When upstream fallback is enabled, [libraries that aren't built by Chainguard](#packages-not-available-in-chainguard-libraries) are subject to Chainguard's security controls.

{{< tabs >}}

{{% tab title="pip and uv" %}}

```shell
chainctl libraries verify --detailed ./.venv/
```

{{% /tab %}}

{{% tab title="Poetry" %}}

Poetry's `virtualenv` location depends on your configuration. If `virtualenvs.in-project` is enabled, it's `.venv` in the project directory. Otherwise, find it with:

```bash
poetry env info --path
```

Then:

```bash
chainctl libraries verify --detailed $(poetry env info --path)
```

{{% /tab %}}

{{< /tabs >}}

A successful result shows what percentage of your project's dependencies were built by Chainguard.

> For full details on verification options and output, see [Verification: Analyze Python packages](/chainguard/libraries/verification/#analyze-python-packages).

Reviewer 3: ```suggestion
For full details on verification options and output, check out [Verification: Analyze Python packages](/chainguard/libraries/verification/#analyze-python-packages).
```

## Step 6: Commit and roll out

Commit the updated lockfile and any non-sensitive configuration changes. Apply the same index, cache, and hash-update steps to other developer workstations and build servers as you migrate them, including CI/CD platforms and any infrastructure that builds the application or installs dependencies.

> For organization-wide rollout using a repository manager, see the [global configuration documentation](/chainguard/libraries/python/global-configuration/).

Reviewer 3: ```suggestion
For organization-wide rollout using a repository manager, refer to the [global configuration documentation](/chainguard/libraries/python/global-configuration/).
```

> ## Packages not available in Chainguard Libraries

Reviewer 3: would this make sense in the troubleshooting section? I wonder if it stands out weirdly on its own here

```shell
chainctl libraries entitlements list
```

For repository manager setups, Chainguard recommends using the configurable fallback rather than configuring a separate public registry fallback in your repository manager, to preserve Chainguard’s security controls.

Learn more about upstream fallback configurations in the [Libraries Overview](/chainguard/libraries/overview/#upstream-fallback-and-controls).

## Troubleshooting

### "403 Forbidden" from the index despite a working keyring

Check for an old or incorrectly scoped `.netrc` entry for `libraries.cgr.dev`, which silently takes priority over keyring authentication. Run with full verbosity to confirm the credential source:

```shell
pip install -vvv <package> 2>&1 | grep -B2 -A2 "credentials\|40[13]"
```

### A pinned dependency shows "not verified" after running `chainctl libraries verify`

Check your install output for a local wheel build (`Building wheel for <package>`), which indicates Chainguard has only a source distribution for that specific pinned version.

## Next steps

> * To apply this configuration across your whole organization using a repository manager, see the [global configuration](/chainguard/libraries/python/global-configuration/) documentation.  

Reviewer 3: ```suggestion
* To apply this configuration across your whole organization using a repository manager, check out the [global configuration](/chainguard/libraries/python/global-configuration/) documentation.  
```

> * To verify downloaded packages were built by Chainguard, see the [verification](/chainguard/libraries/verification/) documentation.  

Reviewer 3: ```suggestion
* To verify downloaded packages were built by Chainguard, refer to the [verification](/chainguard/libraries/verification/) documentation.  
```

> * For full per-tool configuration reference, see the [build configuration](/chainguard/libraries/python/build-configuration/) documentation.

Reviewer 3: ```suggestion
* For full per-tool configuration reference, refer to the [build configuration](/chainguard/libraries/python/build-configuration/) documentation.
```
